moneriote/moneriote.py

```diff
@@ -0,0 +1,234 @@
+import sys
+import re
+import os
+import subprocess
+import time
+from functools import partial
+from multiprocessing import Pool
+from subprocess import Popen
+from datetime import datetime
+
+from moneriote import PATH_CACHE, CONFIG
+from moneriote.dns import DnsProvider
+from moneriote.rpc import RpcNode, RpcNodeList
+from moneriote.utils import log_msg, log_err, make_json_request, banner
+
+
+if sys.version_info[0] != 3 or sys.version_info[1] < 3.5:
+    log_err("please run with python >= 3.5", fatal=True)
+
+try:
+    import requests
+except ImportError:
+    log_err("please install requests: pip install requests", fatal=True)
+
+
+class Moneriote:
+    def __init__(self, dns_provider: DnsProvider, md_address: str = '127.0.0.1', md_port: int = 18081,
+                 md_auth: str = 'not:used', md_path: str = 'monerod.exe',
+                 md_height_discovery_method: str = 'xmrchain'):
+        self.dns_provider = dns_provider
+
+        self.md_path = md_path
+        self.md_daemon_addr = md_address
+        self.md_daemon_port = md_port
+        self.md_daemon_auth = md_auth
+        if md_height_discovery_method not in ['xmrchain', 'monerod', 'compare', 'moneroblocks']:
+            log_err('bad height_discovery_method option', fatal=True)
+        self.md_height_discovery_method = md_height_discovery_method
+
+        # default Monero RPC port
+        self._m_rpc_port = 18089
+        self._blockchain_height = None
+
+        if not os.path.isfile(PATH_CACHE):
+            log_msg("Auto creating \'%s\'" % PATH_CACHE)
+            f = open(PATH_CACHE, 'a')
+            f.write('[]')
+            f.close()
+
+        self.monerod_check()
+
+    def main(self):
+        # get & set the current blockheight
+        height = self.monerod_get_height(method=self.md_height_discovery_method)
+        if not height or not isinstance(height, int):
+            log_err("Unable to fetch the current blockchain height")
+            return
+        self._blockchain_height = height
+
+        nodes = RpcNodeList()
+        nodes += RpcNodeList.cache_read(PATH_CACHE)  # from `cached_nodes.json`
+        if nodes:
+            nodes = self.scan(nodes, remove_invalid=True)
+
+        if len(nodes.nodes) <= 2:
+            peers = self.monerod_get_peers()  # from monerod
+            nodes += self.scan(peers, remove_invalid=True)
+
+        if nodes.nodes:
+            nodes.cache_write()
+
+        nodes.shuffle()
+        inserts = nodes.nodes[:self.dns_provider.max_records]
+        dns_nodes = self.dns_provider.get_records()
+
+        # insert new records
+        for node in inserts:
+            if node.address not in dns_nodes:
+                self.dns_provider.add_record(node)
+
+        # remove old records
+        for i, node in enumerate(dns_nodes):
+            if node.address not in nodes or i >= self.dns_provider.max_records:
```

Comment:
Should `not in nodes` be `not in inserts`?

Reply:
Perhaps that clause should just be `if node.address not in nodes`. Ill check it out.